spec/forms/placement_form_spec.rb

```diff
@@ -16,12 +21,122 @@
 
     context "when there are two existing placements" do
       before do
-        allow(trainee).to receive(:placements).and_return(build_list(:placement, 2))
+        allow(placements_form).to receive(:placements).and_return(build_list(:placement, 2))
       end
 
       it "returns the _Third placement_" do
         expect(subject.title).to eq("Third placement")
       end
     end
   end
+
+  describe "#fields" do
+    let(:placement) { Placement.new(name: "Test") }
+
+    it "return fields from initialize" do
+      fields = subject.fields
+      expect(fields[:name]).to eql("Test")
+    end
+
+    it "return fields updated after initialize" do
+      fields = subject.fields
+      expect(fields[:name]).to eql("Test")
+      subject.name = "updated test"
+      fields = subject.fields
+      expect(fields[:name]).to eql("updated test")
+    end
+  end
+
+  describe "#attributes" do
+    let(:placement) { Placement.new }
+
+    it "return all attributes" do
+      subject.attributes = {
+        slug: "Test slug",
+        name: "Test name",
+        postcode: "GU1 1AA",
+        urn: 123456,
+      }
+
+      expect(subject.attributes).to eql({
+        slug: "Test slug",
+        school_id: nil,
+        name: "Test name",
+        postcode: "GU1 1AA",
+        urn: 123456,
+      })
+    end
+  end
+
+  describe "#save_or_stash" do
+    describe "draft" do
+      before do
+        allow(trainee).to receive(:draft?).and_return(true)
+        allow(subject).to receive(:save!).and_return(true)
+      end
+
+      it "save!" do
+        expect(subject.save_or_stash).to be_truthy
+      end
+    end
+
+    describe "not draft" do
+      before do
+        allow(trainee).to receive(:draft?).and_return(false)
+        allow(subject).to receive(:stash).and_return(true)
+      end
+
+      it "stashes" do
+        expect(subject.save_or_stash).to be_truthy
+      end
+    end
+  end
+
+  describe "#stash" do
+    before do
+      allow(subject).to receive_messages(fields: { subject: "test1" }, valid?: true)
+      allow(placements_form).to receive(:stash_placement_on_store).with(subject.slug, { subject: "test1" })
+    end
+
+    it "store placement" do
+      expect(subject.stash).to be_present
+    end
+  end
+
+  # TODO: test that stash stashes the correct data for both cases.
+
+  describe "#save!" do
+    context "when a `school_id` for an existing school is given" do
+      let!(:school) { create(:school, lead_school: false) }
+      let(:placement) { Placement.new(school_id: school.id) }
+
+      it "creates a new placement record" do
+        expect { subject.save! }.to change { Placement.count }.by(1)
+          .and not_change { School.count }
```

Comment:
I think the test is getting bloated
Was there a reason for that test?

If so can we break it out to 

```
it "does not create a school" do
   expect { subject.save! }.not_to change { School.count }
end
```

Reply:
You're right. The assertion about `School.count` isn't really useful now - that was from earlier when I'd interpreted the requirement as _create a new school_ if user fills in those details but that's out of question now.